File: atlas/examples/kotlin/lib/src/main/kotlin/dev/atlas/example/lib/ReadingList.kt
package dev.atlas.example.lib

import com.google.common.base.CharMatcher
import org.apache.commons.lang3.StringUtils

/**
 * Creates and retains reading-list items behind a reusable library API.
 */
class ReadingList {
    private val items = mutableListOf<ReadingListItem>()

    /**
     * Adds one normalized title to the list.
     *
     * @param title User-supplied book title containing non-whitespace text.
     * @return Created immutable reading-list item.
     */
    fun add(title: String): ReadingListItem {
        val normalized = StringUtils.normalizeSpace(title).trim()
        require(normalized.isNotEmpty()) { "A book title is required." }
        val slug = CharMatcher.whitespace().trimAndCollapseFrom(normalized.lowercase(), '-')
            .replace(Regex("[^a-z0-9-]"), "")
        val item = ReadingListItem(normalized, slug)
        this.items.add(item)
        return item
    }
}

File: atlas/examples/typescript/packages/lib/src/ReadingList.ts
import { z } from "zod";

import { ReadingListItem } from "./ReadingListItem.js";

/** Validates the library's bounded in-memory item count. */
const readingListCapacitySchema = z.number().int().positive().max(100);

/**
 * Creates and retains reading-list items behind a small reusable library API.
 */
export class ReadingList {
  readonly #items: ReadingListItem[] = [];

  /**
   * Creates an empty reading list with a bounded capacity.
   *
   * @param capacity - Maximum item count from one through one hundred.
   */
  public constructor(private readonly capacity: number = 100) {
    readingListCapacitySchema.parse(capacity);
  }

  /**
   * Adds a normalized title to the list.
   *
   * @param title - User-supplied book title.
   * @returns The created reading-list item.
   */
  public add(title: string): ReadingListItem {
    if (this.#items.length >= this.capacity)
      throw new Error("Reading list capacity reached.");
    const item = ReadingListItem.create(title);
    this.#items.push(item);
    return item;
  }
}
